zxlive/commands.py

```diff
@@ -238,3 +244,92 @@ def toggle(self) -> None:
         self.update_graph_view()
 
     undo = redo = toggle
+
+
+@dataclass
+class AddRewriteStep(SetGraph):
+    """Adds a new rewrite to the proof.
+
+    The rewrite is inserted after the currently selected step. In particular, it
+    replaces all rewrites that were previously after the current selection.
+    """
+    step_view: QListView
+    name: str
+    diff: Optional[GraphDiff] = None
+
+    _old_selected: Optional[int] = field(default=None, init=False)
+    _old_steps: list[tuple[Rewrite, GraphT]] = field(default_factory=list, init=False)
+
+    @property
+    def proof_model(self) -> ProofModel:
+        model = self.step_view.model()
+        assert isinstance(model, ProofModel)
+        return model
+
+    def redo(self) -> None:
+        # Remove steps from the proof model until we're at the currently selected step
+        self._old_selected = self.step_view.currentIndex().row()
+        self._old_steps = []
+        for _ in range(self.proof_model.rowCount() - self._old_selected - 1):
+            self._old_steps.append(self.proof_model.pop_rewrite())
+
+        diff = self.diff or GraphDiff(self.g, self.new_g)
+        self.proof_model.add_rewrite(Rewrite(self.name, diff), self.new_g)
+
+        # Select the added vertex
+        idx = self.step_view.model().index(self.proof_model.rowCount() - 1, 0, QModelIndex())
+        self.step_view.selectionModel().blockSignals(True)
+        self.step_view.setCurrentIndex(idx)
+        self.step_view.selectionModel().blockSignals(False)
+        super().redo()
+
+    def undo(self) -> None:
+        # Undo the rewrite
+        self.proof_model.pop_rewrite()
+
+        # Add back steps that were previously removed
+        for rewrite, graph in reversed(self._old_steps):
+            self.proof_model.add_rewrite(rewrite, graph)
+
+        # Select the previously selected step
+        assert self._old_selected is not None
+        self.step_view.selectionModel().blockSignals(True)
+        idx = self.step_view.model().index(self._old_selected, 0, QModelIndex())
+        self.step_view.setCurrentIndex(idx)
+        self.step_view.selectionModel().blockSignals(False)
+        super().undo()
+
+
+@dataclass
+class GoToRewriteStep(SetGraph):
+    """Shows the graph at some step in the proof.
+
+    Undoing returns to the previously selected proof step.
+    """
+
+    def __init__(self, graph_view: GraphView, step_view: QListView, old_step: int, step: int) -> None:
```

Comment:
I thought there was some problem with redefining the __init__ in a dataclass?

Reply:
We'll run into issues once we try to subclass `GoToRewriteStep` but this works fine for now